### ISSUE TITLE
Memory Model (ordering + reads from = happens before) : behavior of fence

### ISSUE THREAD

**Issue by anshumang** [openshmem-org/specification#232]
Umbrella issue #229
V : will order
X : will not order
- : ordered
                              same PE  |  different PE
p/put/iput          |      V          |        X
put_nbi               |      V          |        X
g/get/iget           |      V          |        V
get_nbi                |      V          |        X
non-fetch AMOs  |     V          |        X
fetch AMOs          |     -           |        -

**Comment by jdinan** [comment]
Can you add some prose to help us understand what's posted here?  I assume that same/different PE means the given operation class when two are issued by the same PE with an intervening fence, versus when two are issued by different PEs?  Where and when is the fence performed in the latter case?  Also, how does O relate to fence?

**Comment by anshumang** [comment]
@jdinan I have added text to clarify. Is the table still hard to understand?

**Comment by jdinan** [comment]
Do operations target the same or different memory locations?  Also, isn't ordering of load instructions architecture dependent?

**Comment by anshumang** [comment]
@jdinan I have updated the text to indicate that the target location can be different. Let me know if it is not clear.

> Also, isn't ordering of load instructions architecture dependent?

But should `shmem_fence` not enforce ordering of memory loads on memory obtained by `shmem_ptr`?

**Comment by anshumang** [comment]
I agree that `memory load` should be `V` but it is missing in OpenSHMEM v1.4 spec -
`Assures ordering of delivery of Put, AMO, memory store, and nonblocking Put routines to symmetric data objects`. In absence of `memory load` above, I chose `O` with this interpretation of v1.4 spec - ordered because `shmem_fence` is not required.

**Comment by jdinan** [comment]
IIRC, loads are currently ordered only by `shmem_wait` and possibly `shmem_barrier`/`sync`.  My gut feeling is that `shmem_fence` should not guarantee load ordering since it doesn't ensure that updates are present and visible in memory.

**Comment by naveen-rn** [comment]
I suppose this is not for OpenSHMEM-1.5.
